dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_kernel.cu

```diff
@@ -0,0 +1,131 @@
+#include "dali/core/static_switch.h"
+
+namespace dali {
+namespace kernels {
+namespace jpeg {
+
+KernelRequirements JpegDistortionBaseGPU::Setup(KernelContext &ctx,
+                                                const TensorListShape<3> &in_shape,
+                                                bool horz_subsample, bool vert_subsample) {
+  horz_subsample_ = horz_subsample;
+  vert_subsample_ = vert_subsample;
+  KernelRequirements req;
+  ScratchpadEstimator se;
+  int nsamples = in_shape.num_samples();
+  se.add<SampleDesc>(AllocType::Host, nsamples);
+  se.add<SampleDesc>(AllocType::GPU, nsamples);
+
+  chroma_shape_.resize(nsamples);
+  for (int i = 0; i < nsamples; i++) {
+    auto chroma_sh = chroma_shape_.tensor_shape_span(i);
+    auto sh = in_shape.tensor_shape_span(i);
+    // used to generate logical blocks (one thread per chroma pixel)
+    chroma_sh[0] = div_ceil(sh[0], 1 + vert_subsample_);
+    chroma_sh[1] = div_ceil(sh[1], 1 + horz_subsample_);
+  }
+
+  block_setup_.SetBlockDim(dim3(32, 16, 1));
+  int xblock = 64*(2-horz_subsample_);
+  int yblock = 128;
+  block_setup_.SetDefaultBlockSize({xblock, yblock});
+  block_setup_.SetupBlocks(chroma_shape_, true);
+  int nblocks = block_setup_.Blocks().size();
+  se.add<BlockDesc>(AllocType::GPU, nblocks);
+
+  req.scratch_sizes = se.sizes;
+  req.output_shapes = {in_shape};
+  return req;
+}
+
+void JpegDistortionBaseGPU::SetupSampleDescs(const OutListGPU<uint8_t, 3> &out,
+                                             const InListGPU<uint8_t, 3> &in,
+                                             span<const int> quality) {
+  const auto &in_shape = in.shape;
+  int nsamples = in_shape.num_samples();
+  sample_descs_.resize(nsamples);
+  for (int i = 0; i < nsamples; i++) {
+    auto &sample_desc = sample_descs_[i];
+    auto in_sh = in_shape.tensor_shape_span(i);
+    auto width = in_sh[1];
+    auto height = in_sh[0];
+    sample_desc.in = in[i].data;
+    sample_desc.out = out[i].data;
+    sample_desc.size.x = width;
+    sample_desc.size.y = height;
+    sample_desc.strides.x = 3;
+    sample_desc.strides.y = width * 3;
+    int q;
+    if (quality_.empty()) {
+      q = 95;
+    } else if (quality_.size() == 1) {
+      q = quality_[0];
+    } else {
+      q = quality_[i];
+    }
+    sample_desc.luma_Q_table = GetLumaQuantizationTable(q);
+    sample_desc.chroma_Q_table = GetChromaQuantizationTable(q);
+  }
+}
+
+void JpegCompressionDistortionGPU::Run(KernelContext &ctx, const OutListGPU<uint8_t, 3> &out,
+                                       const InListGPU<uint8_t, 3> &in, span<const int> quality) {
+  const auto &in_shape = in.shape;
+  int nsamples = in_shape.num_samples();
+  if (quality.size() > 1 && quality.size() != nsamples) {
+    throw std::invalid_argument(make_string("Received ", quality.size(),
+                                            " quality values but the batch contains ", nsamples,
+                                            " samples."));
```

Comment:
How about pointing out in the error message, that the number of quality values and batch size have to be the same? It's apparent from code, but I feel it's not obvious reading the err msg only